### ISSUE TITLE
New netlink specific iter function checking `Nlmsg::Done`

### ISSUE THREAD

**Issue by jbaublitz** [jbaublitz/neli#30]
Add another `NetlinkSocket` iter function that checks `NlmF::Multi` and ends when `Nlmsg::Done` is the message type.

**Comment by skrap** [comment]
I am using a possible implementation of this, which I'll paste in here.  It's a bit wordy because the Option<Result<_,_>> type is not compatible with the `?` operator, so I had to write out some of the error handling by hand.  Feel free to adapt, ignore, copy, whatever :)

```
struct NlMultiIter<'a, T, P> {
    socket: &'a mut neli::socket::NlSocket,
    buffer: Vec<u8>,
    position: usize,
    data_type: PhantomData<T>,
    data_payload: PhantomData<P>,
}

impl<'a, T, P> NlMultiIter<'a, T, P> {
    pub fn new(socket_ref: &'a mut NlSocket) -> Self {
        NlMultiIter { socket: socket_ref, buffer: vec![], position: 0, data_type: PhantomData, data_payload: PhantomData }
    }

    fn clear(&mut self) {
        self.position = 0;
        self.buffer.clear();
    }
}

impl<'a, T, P> Iterator for NlMultiIter<'a, T, P> where T: Nl + NlType, P: Nl {
    type Item = Result<Nlmsghdr<T, P>, NlError>;

    fn next(&mut self) -> Option<Result<Nlmsghdr<T, P>, NlError>> {
        // fill buffer if necessary.
        if self.position + 1 >= self.buffer.len() {
            self.buffer = vec![0; MAX_NL_LENGTH];
            match self.socket.recv(&mut self.buffer, 0) {
                Ok(0) => return Some(Err(NlError::new("No data could be read from the socket"))),
                Err(e) => return Some(Err(e.into())),
                Ok(cnt) => self.buffer.truncate(cnt as usize),
            }
            self.position = 0;
        }
        
        // read packet length
        let len = match u32::deserialize(&mut StreamReadBuffer::new(&self.buffer[self.position..])) {
            Err(e) => return Some(Err(e.into())),
            Ok(len) => len as usize,
        };
        if len + self.position > self.buffer.len() {
            let err = Err(NlError::new(&format!("Incomplete packet of {} bytes, only {} bytes in buffer", len, self.buffer.len() - self.position)));
            self.clear();
            return Some(err);
        }

        // frame packet as slice
        // check if packet is "Done"
        let mut packet_reader = StreamReadBuffer::new(&self.buffer[self.position..self.position+len]);
        match Nlmsghdr::<Nlmsg, NlEmpty>::deserialize(&mut packet_reader) {
            Err(e) => {
                self.position += len;
                return Some(Err(e.into()));
            },
            Ok(nl) => match nl.nl_type {
                Nlmsg::Done => {
                    self.position += len;
                    return None
                },

                Nlmsg::Error => {
                    self.position += len;
                    return Some(Err(NlError::new("rtnetlink error.")));
                }

                _ => {}
            },
        }

        // try reading packet as T,P
        let mut packet_reader = StreamReadBuffer::new(&self.buffer[self.position..self.position+len]);
        let nlmsg = Nlmsghdr::<T,P>::deserialize(&mut packet_reader);
        self.position += len;
        match nlmsg {
            Err(e) => {
                return Some(Err(e.into()));
            }
            Ok(msg) => Some(Ok(msg))
        }
    }
}```

**Comment by jbaublitz** [comment]
This is now able to be toggled on or off in the `.iter()` function in the branch `v0.5.0-dev`. Closing.
